src/query/storage/m3/consolidators/multi_fetch_result.go

```diff
@@ -114,48 +136,70 @@ func (r *multiResult) FinalResult() (SeriesFetchResult, error) {
 	r.Lock()
 	defer r.Unlock()
 
-	result := SeriesFetchResult{Metadata: r.metadata}
+	// result := SeriesFetchResult{Metadata: r.metadata}
 	err := r.err.LastError()
 	if err != nil {
-		return result, err
+		return NewEmptyFetchResult(r.metadata), err
 	}
 
 	if r.mergedIterators != nil {
-		result.SeriesIterators = r.mergedIterators
-		return result, nil
+		return NewSeriesFetchResult(r.mergedIterators, nil, r.metadata)
 	}
 
 	if len(r.seenIters) == 0 {
-		result.SeriesIterators = encoding.EmptySeriesIterators
-		return result, nil
+		return NewSeriesFetchResult(encoding.EmptySeriesIterators, nil, r.metadata)
 	}
 
 	// can short-cicuit in this case
 	if len(r.seenIters) == 1 {
-		result.SeriesIterators = r.seenIters[0]
-		return result, nil
+		return NewSeriesFetchResult(r.seenIters[0], nil, r.metadata)
 	}
 
 	// otherwise have to create a new seriesiters
-	numSeries := len(r.dedupeMap)
+	dedupedList := r.dedupeMap.list()
+	numSeries := len(dedupedList)
 	r.mergedIterators = r.pools.MutableSeriesIterators().Get(numSeries)
 	r.mergedIterators.Reset(numSeries)
+	if r.mergedTags == nil {
+		r.mergedTags = make([]models.Tags, numSeries)
+	}
+
+	lenCurr, lenNext := len(r.mergedTags), len(dedupedList)
+	if lenCurr < lenNext {
+		// If incoming list is longer, expand the stored list.
+		r.mergedTags = append(r.mergedTags, make([]models.Tags, lenNext-lenCurr)...)
+	} else if lenCurr > lenNext {
+		// If incoming list somehow shorter, shrink stored list.
+		r.mergedTags = r.mergedTags[:lenNext]
+	}
 
-	i := 0
-	for _, res := range r.dedupeMap {
+	for i, res := range dedupedList {
 		r.mergedIterators.SetAt(i, res.iter)
-		i++
+		r.mergedTags[i] = res.tags
```

Comment:
Do we actually use `r.mergedTags` anywhere? I see it constructed but I only ever see us returning r.mergedIterators and r.metadata?

Reply:
Yeah, looks like it was a bug, would cause tags to be recalculated each time
nice catch!